### PR TITLE
Issue #5951: Add indentation Support for Annotation Array Initialisation

### PR DESCRIPTION
**Pull request by Abhishek-kumar09**
Fixes Issue #5951, #8980  

https://github.com/pmd/pmd/pull/2925 - merged
https://github.com/pgjdbc/pgjdbc/pull/1961

Diff Regression projects: https://gist.githubusercontent.com/Abhishek-kumar09/0745ecf11934e0214b9187f9d4896e4a/raw/7994fb09f0b5d073f1937e62faecdd5cb3908e64/project-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Abhishek-kumar09/0745ecf11934e0214b9187f9d4896e4a/raw/2143ba7f21dcafb998019a1ca2c0ee3f1c510646/my_check.xml

### CODE REVIEW

**Comment by rnveach** [comment]
CI must pass before the PR is reviewed. Regression will also have to be provided for any changes to production code.

**Comment by Abhishek-kumar09** [comment]
> CI must pass before the PR is reviewed.

OK, I will do it.
> Regression will also have to be provided for any changes to production code.

But, RegressionTests are yet not functional for Indentation until Issue #7734 is resolved.

**Comment by davebarda** [comment]
Any update? waiting for it...

**Comment by romani** [comment]
@Abhishek-kumar09, please finish this PR before creation of new PRs

if you do not want to continue it, close it.

**Comment by Abhishek-kumar09** [comment]
Sorry @romani I lost the track of it.
A small help, I don't have any previous knowledge of using array annotations with java.
```
@T.Foo({
  @T.Bar, // line 2
})
class T {

  @interface Foo {
    Bar[] value() default {};
  }

  @interface Bar {}
}
```
Here what should line 2 follow, `basicOffset` or `arrayInitIndent`

**Comment by Abhishek-kumar09** [comment]
The pmd config is: 
```
    <module name="Indentation">
      <property name="severity" value="error"/>
      <property name="caseIndent" value="0"/>
    </module>
```
The ci is failing is failing at pmd config because the children is indented at level 8, which should not be because it is neither the `arrayInitIndent` nor `BasicOffset` or `lineWrapping`. So that test is failing.
https://checkstyle.semaphoreci.com/jobs/7c9424c6-1610-40a0-ae1d-9a5bcae65dcf#L2074

Currently, The children of Annotation Array Init follows 4 indent Levels:
1. the indent of its parent
2. The basicOffset
3. The lineWrapping
4. The arrayInitIndent
Any indent that follows any of these would be good.

**Comment by romani** [comment]
>what should line 2 follow, basicOffset or arrayInitIndent

It should follow arrayInitIndent , it is not a block, as in block separator is ";", but in array it is ",".

**Comment by romani** [comment]
@Abhishek-kumar09 , to fix semaphore violations (pmd) , you need to fork pmd, make changes to follow new indentation rules, we need to send PR to pmd project to be merged. As you share link to PR I can show you how update semaphore to use your fork for validation.
It is at https://github.com/checkstyle/checkstyle/blob/5a5cc1bcea1fe9f763c5a416de3f88106e3cb055/.ci/validation.sh#L87
If you change link to repo https://github.com/checkstyle/checkstyle/blob/5a5cc1bcea1fe9f763c5a416de3f88106e3cb055/.ci/validation.sh#L93 to your fork and do extra command to checkout to your branch, when semaphore pass.

**Comment by romani** [comment]
Wercker is failing at:
```
> Task :postgresql:checkstyleTest
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:19:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:20:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:21:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:22:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:23:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:24:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]
[ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/pgjdbc/pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/OptionalTestSuite.java:25:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 2, 4. [Indentation]

> Task :postgresql:checkstyleTest FAILED

FAILURE: Build failed with an exception.
```

**Comment by Abhishek-kumar09** [comment]
OK, I will check all the other repos too, and see where the changes need to be done.

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by romani** [comment]
@Abhishek-kumar09 , while PMD PR is in progress of acceptance, you can update semaphore config to use you branch
here is example how we did it for xwiki project.
https://github.com/checkstyle/checkstyle/blob/1e9dd66b3e50f0feefc217264778b9efecb3e965/.ci/wercker.sh#L101-L103

**Comment by romani** [comment]
From wercker

```
Command: ./.ci/wercker.sh no-error-orekit

[INFO] --- maven-checkstyle-plugin:3.1.1:check (default-cli) @ orekit ---
[INFO] There is 1 error reported by Checkstyle 8.38-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/annotation/DefaultDataContext.java:[34,9] (indentation) Indentation: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 4, 9, 12.
[INFO] ------------------------------------------------------------------------
[INFO] 
```

**Comment by romani** [comment]
From wercker
```
ommand: ./.ci/wercker.sh no-error-xwiki

[INFO] --- maven-checkstyle-plugin:3.1.1:check (default) @ xwiki-commons-crypto-cipher ---
[INFO] Starting audit...
[ERROR] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-core/xwiki-commons-crypto/xwiki-commons-crypto-cipher/src/main/java/org/xwiki/crypto/cipher/internal/symmetric/factory/BcRc5b64CbcPaddedCipherFactory.java:42:9: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 4, 21, 23. [Indentation]
[WARN] /pipeline/source/.ci-temp/xwiki-commons/xwiki-commons-core/xwiki-commons-crypto/xwiki-commons-crypto-cipher/src/main/java/org/xwiki/crypto/cipher/internal/asymmetric/factory/AbstractBcAsymmetricCipherFactory.java:146:11: Comment matches to-do format 'TODO'. [TodoComment]
Audit done.
[INFO] There is 1 error reported by Checkstyle 8.38-SNAPSHOT with checkstyle.xml ruleset.
[ERROR] src/main/java/org/xwiki/crypto/cipher/internal/symmetric/factory/BcRc5b64CbcPaddedCipherFactory.java:[42,9] (indentation) Indentation: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 4, 21, 23.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for XWiki Commons - Parent POM 12.7-SNAPSHOT:
[INFO] 
```

**Comment by Abhishek-kumar09** [comment]
Line: 9
[ERROR] src/main/java/org/xwiki/crypto/cipher/internal/symmetric/factory/BcRc5b64CbcPaddedCipherFactory.java:[42,9] (indentation) Indentation: 'annotation array initialization' child has incorrect indentation level 8, expected level should be one of the following: 4, 21, 23.

Which repo, cant get the file in the checkstyle/xwiki

**Comment by romani** [comment]
https://github.com/checkstyle/xwiki-commons/blob/master/xwiki-commons-core/xwiki-commons-crypto/xwiki-commons-crypto-cipher/src/main/java/org/xwiki/crypto/cipher/internal/symmetric/factory/BcRc5b64CbcPaddedCipherFactory.java

**Comment by romani** [comment]
Github, rebase

**Comment by Abhishek-kumar09** [comment]
```
[ivy:cachepath] 		::::::::::::::::::::::::::::::::::::::::::::::01:16
[ivy:cachepath] 		::          UNRESOLVED DEPENDENCIES         ::01:16
[ivy:cachepath] 		::::::::::::::::::::::::::::::::::::::::::::::01:16
[ivy:cachepath] 		:: com.puppycrawl.tools#checkstyle;8.38-SNAPSHOT: not found01:16
[ivy:cachepath] 		::::::::::::::::::::::::::::::::::::::::::::::
```

**Comment by romani** [comment]
Strange failure ... restarted.

File: .ci/validation.sh
Comment (romani) [inline .ci/validation.sh:93]: https://github.com/pmd/pmd/pull/2925 is merged, please restore original repo.

File: .ci/wercker.sh
Comment (romani) [inline .ci/wercker.sh:85]: please send PR https://github.com/CS-SI/Orekit/pulls/Abhishek-kumar09

File: .ci/wercker.sh
Comment (romani) [inline .ci/wercker.sh:101]: https://github.com/xwiki/xwiki-commons/pulls/Abhishek-kumar09

please send PR to them.

send PR to our repo - https://github.com/checkstyle/xwiki-commons/tree/5951-indentation

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [inline .ci/wercker.sh:67]: REL 42.2.16 is not a branch.
Moreover, I had sent PR to master branch, did I made a mistake :(

File: .ci/validation.sh
Comment (Abhishek-kumar09) [reply .ci/validation.sh]: Done

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: Done fix is merged.
https://github.com/CS-SI/Orekit/pull/28

https://github.com/checkstyle/checkstyle/pull/8083/files#diff-8b07316a80f4b78a810d359a8be767cb13f5831e827aaa3b64cc36b6d133c2a0

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: Done 
https://github.com/checkstyle/xwiki-commons/pull/1

File: .ci/wercker.sh
Comment (romani) [reply .ci/wercker.sh]: No mistake, release is set in stone.
Usage of master is good if sources are build able, but their CI is red.

File: .ci/wercker.sh
Comment (romani) [reply .ci/wercker.sh]: Merged.
Please update to use our fork.

Please send PR to xwiki project to let them quickly merge it after we release your fix.

File: .ci/wercker.sh
Comment (romani) [reply .ci/wercker.sh]: Ok, please their repo and commit where it was merged ( if we used strict commit)

Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: Already done.

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: > Please send PR to xwiki project to let them quickly merge it after we release your fix.

I sent it yesterday, and they had merged. 
https://github.com/xwiki/xwiki-commons/pull/117

> Please update to use our fork.

done

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: Already done.

Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: You can checkout ci changes here: https://github.com/checkstyle/checkstyle/pull/8083/commits/33189c8557120cb254f64d8d4360c98621dffbf8

File: .ci/wercker.sh
Comment (romani) [inline .ci/wercker.sh:67]: > bdf

https://travis-ci.org/github/checkstyle/checkstyle/jobs/745479736#L306

please do hack
`"51f18bf7755b97a4ab3e560431a3578d119a2b""df"` to avoid extension of spellchecker with bad words

**Review by romani — CHANGES_REQUESTED** [review]
from semaphore ` .ci/validation.sh no-violation-test-josm`:

> `module not found: com.puppycrawl.tools#checkstyle;8.38-SNAPSHOT`

it is annoying and not related 



items:

**Comment by romani** [comment]
Github, rebase

File: .ci/wercker.sh
Comment (Abhishek-kumar09) [reply .ci/wercker.sh]: Its checking spelling when there are consecutive three letters. Nice hack.
Done.

**Comment by Abhishek-kumar09** [comment]
space is needed after comment
https://github.com/checkstyle/checkstyle/blob/master/.semaphore/semaphore.yml#L56-L57

**Comment by romani** [comment]
Github, rebase

**Review by romani — APPROVED** [review]
@Abhishek-kumar09 , thank a lot for you work.
we need one more review from another person, as I am become blind with your code.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationDifficultAnnotations.java
Comment (strkkk) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationDifficultAnnotations.java:41]: `23,25`
I do not have a good knowledge of this check, can you shortly describe why there are such big numbers for expected indentation?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationOddLineWrappingAndArrayInit.java
Comment (strkkk) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationOddLineWrappingAndArrayInit.java:25]: `exp:11,17,53,54`
same here, indentation 53 or 54 sounds a bit weird

**Review by strkkk — COMMENTED** [review]
Just a minor question

@Abhishek-kumar09 also I want to say that your work on indentation issues is very helpful

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationDifficultAnnotations.java
Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationDifficultAnnotations.java:41]: The children are accepting indent levels like arrayInitIndent, lineWrapping and basicOffset.
Those big numbers are because it is also accepting indentLevel from firstNonBlank line after lcurly (Initialization block's) inside which those children are placed.

`23, 25`:
because indentation comment is the first non blank line on after the lcurly. And it is placed at 23 indent level. so 23 and 23+arrayInitIndent(2) = 25.

This is important/helpful when we have code style like this: (where we are indenting other children relative to previous ones) 
```
@Annotation({ @Child1,
              @Child2, 
              @Child3
})
``` 

So yes, if there would not have those comments there then we would not have got `23,25`.

**Comment by Abhishek-kumar09** [comment]
> @Abhishek-kumar09 also I want to say that your work on indentation issues is very helpful

Thank you very much, It made my day when I see those comments from mentors and seniors like you.  Truly before checkstyle I had seen and tried to get started with 20+ open source projects, and always fail and get frustrated with being realized about the thing I don't  understand.
It was because of you and other mentors that I got involved in this community, and wrote some code that gave me enough exposure and most importantly a self-satisfaction that I could now understand few things in such big project. I will be always thankful for being here.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationOddLineWrappingAndArrayInit.java
Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationOddLineWrappingAndArrayInit.java:25]: 53 is the indent level of first comment on line "24".
54 is the lineWrapping + lcurly pos.
Here is the logic:
```
        if (firstChildPos != NOT_EXIST) {
            expectedIndent = IndentLevel.addAcceptable(expectedIndent, firstChildPos, lcurlyPos
                    + getLineWrappingIndentation());
        }
```

Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationOddLineWrappingAndArrayInit.java]: comment was unnecesarily placed too far, corrected.

**Review by strkkk — APPROVED** [review]

**Comment by strkkk** [comment]
@romani I am ok with changes, do you think someone else should review or we can merge after CI pass?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: @romani `forceStrictCondition` is set to `false`. Shouldn't there be no violation here?

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: @Abhishek-kumar09 , can you explain ?
This option allow to use extra spaces if that visually benefitial

Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: Actually it is not following forceStrictCondition.
If it would, then all the Indent-levels after arrayInitIndent/linewrappingIndent  is valid.
Particularly here, the user intent is to align code with 0 lineWrapping and 0 arrayInit indent.

For giving some flexibility over only `arrayInitIndent`, the child is also allowed to have `lineWrapping` and `lcurly+lineWrapping`.

But, yes it feels valid to me now that the child should also follow `forceStrictCondition` being false/true.

What's your take? Should I change the code?

Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java]: @rnveach I have applied the changes. Please have a look.

**Comment by Abhishek-kumar09** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: Continuation of https://github.com/checkstyle/checkstyle/pull/8083#discussion_r530751111 ,

Should 0 really be allowed as an indentation? 
Even google style examples show some indentation in the example provided.
https://google.github.io/styleguide/javaguide.html#s4.8.3.1-array-initializers

**Comment by romani** [comment]
@Abhishek-kumar09 ,

from wercker
> CS_version: 8.39-SNAPSHOT
Cloning into 'pgjdbc'...
fatal: reference is not a tree: 51f18bf7755b97a4ab3e560431a3578d119a2bdf

Looks like we need to come back to your fork and custom branch, or find new SHA for commit that contains your update in pgjdbc.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: @rnveach 
Actually it is checking against config where arrayInit = 0, lineWrapping = 0. 
and thus allowing 0 as an indentation level.

Since google has 2 as arrayInit, and 4 as lineWrapping. It will not allow the same.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java]: sorry for being late on this discussion .... I needed some time to meditate on this.

https://checkstyle.org/config_misc.html#Indentation
from documentation

> forceStrictCondition | Force strict indent level in line wrapping case. If value is true, line wrap                   indent have to be same as lineWrappingIndentation parameter. If value is false,                   line wrap indent could be bigger on any value user would like.

it is clearly not for `arrayInitIndent`. arrayInitIndent is somehow in between of basicOffset and lineWrappingIndentation. Because technically array should be in one line and we just wrap array by `,` in multiple line.
But the same is true for basicOffset, Class/block can be single line `class A { int i; int j; }` and we just wrap it by `;` in separate lines. But we have special property for this - basicOffset. So wrapping by separator between completely independent items should follow some `offset`.

Lets come back to array.
Array in very similar to blocks, as all its inner elements are independent, but not separated by `;` so it has special property.
`lineWrappingIndentation` is more applicable for wrapping inside statement - it wrapping in level deeper than by `;`. For arrays - it is for wrapping inside array element, not between them.

@rnveach , @Abhishek-kumar09 if you both agree ...... Please revert your implementation of `lineWrappingIndentation` for arrays. Sorry for being late.

**Comment by Abhishek-kumar09** [comment]
> Looks like we need to come back to your fork and custom branch, or find new SHA for commit that contains your update in pgjdbc.

yes, made changes...Since according to new changes the previous code that was written is also valid and no need to change the branch or anything. But still I have changes the wercker.sh to checkout the branch 'release/master' which was updated 8 months ago.

**Comment by Abhishek-kumar09** [comment]
@romani Is wercker failure still due to pgjdbc

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: Thanks, sorry, I missed the config.

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java:17]: @Abhishek-kumar09 , 
`0` is not good indentation.
please make code sample like this with `indent:0` and should violate such code.

**Comment by romani** [comment]
yes,

```
Command: ./.ci/wercker.sh no-error-pgjdbc
CS_version: 8.39-SNAPSHOT
Cloning into 'pgjdbc'...
Switched to a new branch 'release/master'
Branch 'release/master' set up to track remote branch 'release/master' from 'origin'.
./.ci/wercker.sh: line 68: ./gradlew: No such file or directory
```

please run it on your local to pass , it is easy `./.ci/wercker.sh no-error-pgjdbc`

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit.java
Comment (Abhishek-kumar09) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit.java:17]: > @Abhishek-kumar09 ,
0 is not good indentation.
please make code sample like this with indent:0 and should violate such code.

Its already here. Indent:0 exp:>=4

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationAnnArrInit2.java]: @rnveach , @Abhishek-kumar09 , please read and if you agree lets proceed in this way.

**Comment by Abhishek-kumar09** [comment]
@rnveach ping

**Review by rnveach — APPROVED** [review]

**Pull request merged**
